### ISSUE TITLE
[cond] cache size limit exceeeded

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#108500]
### 🐛 Describe the bug

Repro

1) Get this stack of branch - https://github.com/pytorch/pytorch/pull/108402
2) Run ` test/functorch/test_control_flow.py::TestControlFlowTraced::test_nested_map_cond_real` or `pytest test/functorch/test_control_flow.py::TestControlFlowTraced::test_cond_nested_with_closure`

Background - I am trying to reduce the cache size limit to 4. Its unclear to me why torch.cond requires even more than one compile. I see many recompiles. Its also counter intuitive because torch.cond is targeted for torch.export.

cc @chauhang @penguinwu @ydwu4 @bdhirsh @bobrenjc93 @aorenste @ezyang @msaroufim @wconstab @zou3519 

### Versions

N/A

**Comment by ydwu4** [comment]
torch.cond now uses torch.compile under the hood. Basically, all cond share the code object. This causes cond reaches the size limit in a single unit test, where we may have nested cond with real tensor, fake tensor inputs and each of them causes a recompilation.

Depending on whether real use-cases consist of multiple cond in a program (beyond cache limit): we could either 1. fix the cond tests by reseting the cache or 2. make cond a special case and allow it to have more caches.

**Comment by ezyang** [comment]
@ydwu4 having all the cond on the same code object will be bad for eager cond perf. @anijain2305 guard trees should help with this no?

**Comment by ydwu4** [comment]
> having all the cond on the same code object will be bad for eager cond perf

Is it bad because of cache look up in dynamo? If we have multiple code objects in cond, we need to provide some kind of caching mechanism in cond to reduce number of re-compilation.

**Comment by anijain2305** [comment]
@ydwu4 Is it possible to compile the cond, true and false branch fuctions separately instead of compiling cond? In that case, the cache will be sitting on the code objects of those functions, instead of `cond`.

**Comment by ydwu4** [comment]
> Is it possible to compile the condition_fn, true and false branch fuctions separately instead of compiling cond

Yeah, that would solve the caching problem. 

But it is essentially the export approach that we gave up previously because the implementation will be pretty hacky and hard to maintain.

**Requirements:**
1. We want to normalize two branches before feeding into torch.ops.higher_order.cond so that branches have same inputs different with closures lifted
2. we can cache the compiled brnaches to avoid repeated re-compilation

**Export approach:**

We'll have to have **a user-level cache** for dynamo compiled function in cond. It 1. stores all the **compiled graphs and wrappers**. Wrappers are needed for normalization requirements listed above. And we need dynamo to inspect the branches to tell us what's the inputs to the wrappers. 2. **can dynamically resolve** the inputs of the wrappers (i.e. guards check success but tensor data is different, we need to resolve and feed new inputs to the cached function). 3. **being able to check guards** to know whether re-compilation are required (may need to re-use the guard checking logic in dynamo)

The other thing is that CondHigherOrderOp logic will be skipped at the top-level cond (nested inner cond is still handled by CondHigherOrderOp) and need to carefully keep them in sync.


@zou3519 @ezyang , any comments on this?

**Comment by zou3519** [comment]
@anijain2305 how urgent is this? (I remember you decided to change the cache size to 8?). Also, out of curiosity, what is the motivation for reducing the cache size?

**Comment by zou3519** [comment]
Talked a bit with Ed, short-term wise we could just figure out how to make the cache limits for cond unlimited, but that's not going to be good for performance. Long-term wise we want something like figuring out how to cache on each cond's callsite (instead of the cond itself), or maybe somehow the true_fn/false_fn

**Comment by jansel** [comment]
My agent was unable to reproduce this on b5f741719293dba49b11165b595d964febaa9ba9. Reopen if you see this on the latest PyTorch nightly
